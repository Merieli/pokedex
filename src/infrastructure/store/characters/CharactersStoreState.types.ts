export interface CharactersStoreState {
    characters: any[];
}
