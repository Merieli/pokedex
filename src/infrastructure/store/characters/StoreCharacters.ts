import { defineStore } from 'pinia';
import { CharactersStoreState } from './CharactersStoreState.types';
import getAllCharacters from './actions/getAllCharacters';

const state = (): CharactersStoreState => ({
    characters: [],
});

const getters = {};

const actions = {
    getAllCharacters,
};

export const useCounterStore = defineStore('characters', {
    state,
    actions,
});
