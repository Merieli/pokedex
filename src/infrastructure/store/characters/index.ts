export * from './CharactersStoreState.types';
export * from './StoreCharacters';
