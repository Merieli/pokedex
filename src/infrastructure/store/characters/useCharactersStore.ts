import { Ref, ref } from 'vue';

import { Character } from '@/domain';
import CharactersGatewayHttp from '@/infrastructure/gateway/CharactersGatewayHttp';

import { defineStore } from 'pinia';

export const useCharactersStore = defineStore('characters', () => {
    const characters: Ref<Character[]> = ref([]);
    const isLoading: Ref<boolean> = ref(false);

    const getAllCharacters = async (): Promise<void> => {
        try {
            isLoading.value = true;
            const api = new CharactersGatewayHttp();
            const response = await api.getAll();
            characters.value = response.results;
        } catch (error) {
            // const { code } = logError('Action getAllCharacters', error);
            // notifyUser(code, 'imprimir todos os personagens')  //função para exibir o erro para o usuário
        } finally {
            isLoading.value = false;
        }
    };

    return {
        characters,
        isLoading,
        getAllCharacters,
    };
});
