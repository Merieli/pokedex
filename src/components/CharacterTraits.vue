<script setup lang="ts">
import { ComputedRef, computed, ref, toRefs } from 'vue';

import { Episode } from '@domain/index';

import { useColorTags } from '@/composables/useColorTags';
import { useSelectedCharacter } from '@/composables/useSelectedCharacter';
import { useCharactersStore } from '@/infrastructure/store/characters';

const store = useCharactersStore();
const { clearSelectedCharacter } = useSelectedCharacter();

const { name, episode, gender, id, image, location, origin, species, status, type } = toRefs(store.selectedCharacter);
// TODO: Location está sumindo ao abrir pela 2° vez validar o porque

const { colorTagSpecie, colorTagStatus } = useColorTags(species.value, status.value || '');

const lastEpisode: ComputedRef<Episode> = computed(() => episode.value[episode.value.length - 1]);

const tab = ref(null);
</script>

<template>
    <Teleport to="body">
        <v-card class="character-traits">
            <v-btn
                class="character-traits__back-button"
                variant="text"
                icon="mdi-arrow-left"
                @click="clearSelectedCharacter"
            >
            </v-btn>
            <v-card-title class="character-traits__title">
                <h3 class="character-traits__name" data-character-traits="name">{{ name }}</h3>
                <h4 class="character-traits__id" data-character-traits="id"># {{ id }}</h4>
            </v-card-title>
            <v-card-subtitle class="character-traits__tags">
                <v-chip
                    class="character-traits__tag"
                    :color="colorTagSpecie"
                    variant="elevated"
                    data-character-traits="specie"
                >
                    {{ species }}
                </v-chip>
                <v-chip
                    v-if="status"
                    class="character-traits__tag"
                    :color="colorTagStatus"
                    variant="elevated"
                    data-character-traits="status"
                >
                    {{ status }}
                </v-chip>
            </v-card-subtitle>

            <v-card-text>
                <img class="character-traits__image" :src="image" alt="" />
            </v-card-text>

            <v-card class="character-traits__sheet">
                <v-tabs v-model="tab" class="character-traits__tabs" color="primary" align-tabs="center">
                    <v-tab value="about">About</v-tab>
                    <v-tab value="episodes">Episodes</v-tab>
                    <v-tab value="clones">Clones</v-tab>
                </v-tabs>

                <v-card-text class="character-traits__sheet-content">
                    <v-window v-model="tab">
                        <v-window-item value="about">
                            <v-list lines="five">
                                <v-row no-gutters>
                                    <v-col cols="5">
                                        <v-list-item title="Gender" :subtitle="gender">
                                            <template #prepend>
                                                <v-icon color="blue">mdi-gender-male-female</v-icon>
                                            </template>
                                        </v-list-item>
                                    </v-col>
                                    <v-col cols="7">
                                        <v-list-item title="Born in" :subtitle="origin">
                                            <template #prepend>
                                                <v-icon color="red">mdi-heart-pulse</v-icon>
                                            </template>
                                        </v-list-item>
                                    </v-col>
                                </v-row>
                                <v-list-item title="Type" :subtitle="type"></v-list-item>
                                <v-list-item title="Location" data-character-traits="location">
                                    <template #prepend>
                                        <v-icon color="green">mdi-earth-box</v-icon>
                                    </template>

                                    <template #subtitle>
                                        <p>Name: {{ location.name }}</p>
                                        <p>Dimension: {{ location.dimension }}</p>
                                        <p>Type: {{ location.type }}</p>
                                    </template>
                                </v-list-item>
                                <v-list-item title="Last episode" data-character-traits="last-episode">
                                    <template #prepend>
                                        <v-icon color="black">mdi-video-box</v-icon>
                                    </template>
                                    <template #subtitle>
                                        <p>Name: {{ lastEpisode.name }}</p>
                                        <p>Season: {{ lastEpisode.season }}</p>
                                        <p>Episode: {{ lastEpisode.number }}</p>
                                    </template>
                                </v-list-item>
                            </v-list>
                        </v-window-item>

                        <v-window-item value="episodes">
                            <v-chip v-for="(currentEpisode, index) in episode" :key="index" class="ma-2">
                                {{ currentEpisode.name }}
                            </v-chip>
                        </v-window-item>

                        <v-window-item value="clones"> </v-window-item>
                    </v-window>
                </v-card-text>
            </v-card>
        </v-card>
    </Teleport>
</template>

<style lang="postcss" scoped>
body {
    @apply overflow-hidden overflow-visible;

    scrollbar-width: none;

    &::-webkit-scrollbar {
        display: none;
    }
}

.character-traits {
    @apply w-full h-full
        fixed top-0
        bg-meri-light;

    &__title {
        @apply flex justify-between items-center;
    }

    &__title,
    &__tags {
        @apply pl-10 pr-10;
    }

    &__tag {
        @apply mr-2 mb-2;
    }

    &__name {
        @apply font-extrabold;
    }

    &__id {
        @apply font-semibold text-base;
    }

    &__back-button {
        @apply ml-8 mt-8;
    }

    &__image {
        @apply rounded-3xl max-w-xs m-auto;
    }

    &__sheet {
        @apply rounded-t-3xl
            w-full max-h-[40%]
            fixed bottom-0;
    }

    &__sheet-content {
        @apply overflow-y-auto;
    }

    &__tabs {
        @apply text-meri-low
            pt-8 h-auto;
    }
}
</style>
<style scoped>
body::-webkit-scrollbar {
    display: none;
}
.v-list-item__content > .v-list-item-subtitle {
    text-transform: capitalize;
}

.character-traits .v-list-item__prepend > .v-icon {
    margin-right: 8px;
}
</style>
