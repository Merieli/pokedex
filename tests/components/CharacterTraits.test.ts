import CharacterTraits from '@components/CharacterTraits.vue';
import { createTestingPinia } from '@pinia/testing';
import { mount } from '@vue/test-utils';
import { Pinia } from 'pinia';
import { createVuetify } from 'vuetify';

import { mockSelectedCharacter } from './__mocks__/mockSelectedCharacter';

import { useCharactersStore } from '@/infrastructure/store/characters';

describe('CharacterTraits.vue', () => {
    let pinia: Pinia;

    const setupWrapper = () => {
        vi.mock('@/infrastructure/gateway/CharactersGatewayHttp');

        pinia = createTestingPinia({
            initialState: {
                characters: {
                    pagination: {
                        currentPage: 1,
                        total: 1,
                        results: 10,
                    },
                    selectedCharacter: mockSelectedCharacter,
                },
            },
        });

        const vuetify = createVuetify();

        return {
            wrapper: mount(CharacterTraits, {
                global: {
                    plugins: [pinia, vuetify],
                    stubs: {
                        teleport: true,
                    },
                },
            }),
            store: useCharactersStore(),
        };
    };

    describe('Integração/Componente', () => {
        describe('👀 Renderização:', () => {
            test('Dado o componente Quando renderizado Então deve possuir html equivalente ao Snapshot gravado', () => {
                const { wrapper } = setupWrapper();

                expect(wrapper.html()).toMatchSnapshot();
            });

            test('Dado uma lista de características do personagem Quando renderizado Então deve exibir o nome e id do personagem', () => {
                const { wrapper } = setupWrapper();
                const characterName = wrapper.find('[data-character-traits="name"]');
                const characterId = wrapper.find('[data-character-traits="id"]');

                expect(characterName.text()).toBe(mockSelectedCharacter.name);
                expect(characterId.text()).toBe(`# ${mockSelectedCharacter.id}`);
            });

            test('Dado uma lista de caraterísticas do personagem Quando renderizado Então deve exibir as tags de espécie e estado de vida', () => {
                const { wrapper } = setupWrapper();
                const characterSpecie = wrapper.find('[data-character-traits="specie"]');
                const characterStatus = wrapper.find('[data-character-traits="status"]');

                expect(characterSpecie.text()).toBe(mockSelectedCharacter.species);
                expect(characterStatus.text()).toBe(mockSelectedCharacter.status);
            });

            test('Dado uma lista de caraterísticas do personagem Quando renderizado Então deve exibir nas abas os dados da localização do personagem', () => {
                const { wrapper } = setupWrapper();
                const location = wrapper.find('[data-character-traits="location"]');

                expect(location.exists()).toBeTruthy();
            });

            test('Dado uma lista de caraterísticas do personagem Quando renderizado Então deve exibir nas abas os dados do último episódio com o personagem', () => {
                const { wrapper } = setupWrapper();
                const location = wrapper.find('[data-character-traits="last-episode"]');

                expect(location.exists()).toBeTruthy();
            });
        });
        // describe('🧠 Comportamento:', ()=> {

        // })
        // describe('🐕 Navegação:', ()=> {

        // })
    });
});
