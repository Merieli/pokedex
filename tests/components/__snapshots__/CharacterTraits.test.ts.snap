// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterTraits.vue > Integra√ß√£o/Componente > üëÄ Renderiza√ß√£o: > Dado o componente Quando renderizado Ent√£o deve possuir html equivalente ao Snapshot gravado 1`] = `
"<teleport-stub to=\\"body\\" data-v-d87f46a6=\\"\\">
  <v-card class=\\"character-traits\\" data-v-d87f46a6=\\"\\">
    <v-btn class=\\"character-traits__back-button\\" variant=\\"text\\" icon=\\"mdi-arrow-left\\" data-v-d87f46a6=\\"\\"></v-btn>
    <v-card-title class=\\"character-traits__title\\" data-v-d87f46a6=\\"\\">
      <h3 class=\\"character-traits__name\\" data-character-traits=\\"name\\" data-v-d87f46a6=\\"\\">Rick Sanchez</h3>
      <h4 class=\\"character-traits__id\\" data-character-traits=\\"id\\" data-v-d87f46a6=\\"\\"># 1</h4>
    </v-card-title>
    <v-card-subtitle class=\\"character-traits__tags\\" data-v-d87f46a6=\\"\\">
      <v-chip class=\\"character-traits__tag\\" color=\\"#f2a38b\\" variant=\\"elevated\\" data-character-traits=\\"specie\\" data-v-d87f46a6=\\"\\">Human</v-chip>
      <v-chip class=\\"character-traits__tag\\" color=\\"green\\" variant=\\"elevated\\" data-character-traits=\\"status\\" data-v-d87f46a6=\\"\\">Alive</v-chip>
    </v-card-subtitle>
    <v-card-text data-v-d87f46a6=\\"\\"><img class=\\"character-traits__image\\" src=\\"https://rickandmortyapi.com/api/character/avatar/1.jpeg\\" alt=\\"\\" data-v-d87f46a6=\\"\\"></v-card-text>
    <v-card class=\\"character-traits__sheet\\" data-v-d87f46a6=\\"\\">
      <v-tabs class=\\"character-traits__tabs\\" color=\\"primary\\" align-tabs=\\"center\\" data-v-d87f46a6=\\"\\">
        <v-tab value=\\"about\\" data-v-d87f46a6=\\"\\">About</v-tab>
        <v-tab value=\\"episodes\\" data-v-d87f46a6=\\"\\">Episodes</v-tab>
        <v-tab value=\\"clones\\" data-v-d87f46a6=\\"\\">Clones</v-tab>
      </v-tabs>
      <v-card-text class=\\"character-traits__sheet-content\\" data-v-d87f46a6=\\"\\">
        <v-window data-v-d87f46a6=\\"\\">
          <v-window-item value=\\"about\\" data-v-d87f46a6=\\"\\">
            <v-list lines=\\"five\\" data-v-d87f46a6=\\"\\">
              <v-row no-gutters=\\"\\" data-v-d87f46a6=\\"\\">
                <v-col cols=\\"5\\" data-v-d87f46a6=\\"\\">
                  <v-list-item title=\\"Gender\\" subtitle=\\"Male\\" data-v-d87f46a6=\\"\\"></v-list-item>
                </v-col>
                <v-col cols=\\"7\\" data-v-d87f46a6=\\"\\">
                  <v-list-item title=\\"Born in\\" subtitle=\\"Earth (C-137)\\" data-v-d87f46a6=\\"\\"></v-list-item>
                </v-col>
              </v-row>
              <v-list-item title=\\"Type\\" subtitle=\\"Human normal\\" data-v-d87f46a6=\\"\\"></v-list-item>
              <v-list-item title=\\"Location\\" data-character-traits=\\"location\\" data-v-d87f46a6=\\"\\"></v-list-item>
              <v-list-item title=\\"Last episode\\" data-character-traits=\\"last-episode\\" data-v-d87f46a6=\\"\\"></v-list-item>
            </v-list>
          </v-window-item>
          <v-window-item value=\\"episodes\\" data-v-d87f46a6=\\"\\">
            <v-chip class=\\"ma-2\\" data-v-d87f46a6=\\"\\">Pilot</v-chip>
            <v-chip class=\\"ma-2\\" data-v-d87f46a6=\\"\\">Lawnmower Dog</v-chip>
            <v-chip class=\\"ma-2\\" data-v-d87f46a6=\\"\\">Anatomy Park</v-chip>
          </v-window-item>
          <v-window-item value=\\"clones\\" data-v-d87f46a6=\\"\\"></v-window-item>
        </v-window>
      </v-card-text>
    </v-card>
  </v-card>
</teleport-stub>"
`;
